### PR TITLE
[kotlin-server] --library=ktor (barebones implementation)

### PR DESCRIPTION
**Pull request by jimschubert**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

This is the starter for Kotlin server generation. Initial library support is ktor.

Opening the PR for feedback and any assistance on fleshing this out is welcome.

See #6147

### CODE REVIEW

**Comment by jimschubert** [comment]
The generator still needs tests, but this adds a basic ktor generator.

There's an issue that still needs to be worked out, in that locations will require all HTTP verbs for a given path to be nested in the same location block. In separate blocks, it results in an exception when authentication is installed in the pipeline (it's seen as a duplicate feature).

**Comment by jimschubert** [comment]
This commit also includes lambdas for simplifying processing in
library-based server generator code. As an example, ktor requires
lowercase http methods while spring (a potential future generator)
would require an uppercase such as `HttpMethod.GET`. It doesn't make sense
to modify these in the operations post-process method because that
format wouldn't be universally desirable.

The lambdas included in the KotlinServerCodegen:
* lowercase: converts all text to lowercase
* uppercase: converts all text to UPPERCASE
* titlecase: converts words (with configurable delim) to Title Case
* indented|indented_8|indented_12|indented_16: these helpers apply the
  same desired indent to all lines of an included fragment's text.

File: bin/windows/kotlin-server-petstore.bat [hunk 1]
```diff
@@ -0,0 +1,10 @@
+set executable=.\modules\swagger-codegen-cli\target\swagger-codegen-cli.jar
+
+If Not Exist %executable% (
+  mvn clean package
+)
+
+REM set JAVA_OPTS=%JAVA_OPTS% -Xmx1024M -DloggerPath=conf/log4j.properties
+set ags=generate  --artifact-id "kotlin-petstore-server" -i modules\swagger-codegen\src\test\resources\2_0\petstore.yaml -l kotlin --library=ktor -o samples\server\petstore\kotlin
```
Comment (wing328) [inline bin/windows/kotlin-server-petstore.bat]: Typo: `-l kotlin` => `-l kotlin-server`

Comment (jimschubert) [reply bin/windows/kotlin-server-petstore.bat]: oops. thanks for catching that. I've made the change.

**Comment by wing328** [comment]
AppVeyor reports the following errors:
```
[00:02:41] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.4:jar (attach-javadocs) on project swagger-codegen: MavenReportException: Error while generating Javadoc:
[00:02:41] [ERROR] Exit code: 1 - C:\projects\swagger-codegen-wh2wu\modules\swagger-codegen\src\main\java\io\swagger\codegen\InlineModelResolver.java:437: warning: no description for @return
[00:02:41] [ERROR] * @return
[00:02:41] [ERROR] ^
[00:02:41] [ERROR] C:\projects\swagger-codegen-wh2wu\modules\swagger-codegen\src\main\java\io\swagger\codegen\languages\ScalaGatlingCodegen.java:199: warning: no description for @param
[00:02:41] [ERROR] * @param swagger
[00:02:41] [ERROR] ^
[00:02:41] [ERROR] C:\projects\swagger-codegen-wh2wu\modules\swagger-codegen\src\main\java\io\swagger\codegen\languages\StaticHtmlGenerator.java:80: warning - Tag @link: can't find processCompiler(Compiler) in io.swagger.codegen.languages.StaticHtmlGenerator
[00:02:41] [ERROR] C:\projects\swagger-codegen-wh2wu\modules\swagger-codegen\src\main\java\io\swagger\codegen\languages\ZendExpressivePathHandlerServerCodegen.java:136: warning: no description for @param
[00:02:41] [ERROR] * @param swagger
[00:02:41] [ERROR] ^
[00:02:41] [ERROR] C:\projects\swagger-codegen-wh2wu\modules\swagger-codegen\src\main\java\io\swagger\codegen\mustache\IndentedLambda.java:23: error: bad use of '>'
[00:02:41] [ERROR] *     {{#indent4}}{{>template}}{{/indent4}}
[00:02:41] [ERROR] ^
[00:02:41] [ERROR] C:\projects\swagger-codegen-wh2wu\modules\swagger-codegen\src\main\java\io\swagger\codegen\mustache\IndentedLambda.java:24: error: bad use of '>'
[00:02:41] [ERROR] *         {{#indent8}}{{>other_template}}{{/indent8}}
[00:02:41] [ERROR] ^
[00:02:41] [ERROR] C:\projects\swagger-codegen-wh2wu\modules\swagger-codegen\src\main\java\io\swagger\codegen\mustache\TitlecaseLambda.java:39: warning: no description for @param
[00:02:41] [ERROR] * @param delimiter
[00:02:41] [ERROR] ^
[00:02:41] [ERROR] 
```
Please take a look when you've time.

**Comment by jimschubert** [comment]
I noticed some errors there, but don't get them locally. Should I fix only the files in this PR or all files listed? I don't get if I changed something that has a negative impact on those unrelated files?

**Comment by wing328** [comment]
Are you using JDK8 locally?

**Comment by jimschubert** [comment]
Ah, no I'm not. I've been using JDK 8 since I got my new machine a few weeks ago. I didn't bother messing around with JDK 7 :]

**Comment by jimschubert** [comment]
Looks like AppVeyor is failing on Java 8, not Java 7:

```
C:\Progra~1\Java\jdk1.8.0\jre\..\bin\javadoc.exe -J-Xmx1024m @options @packages
```

I'm still confused how other PRs are green with the same errors. For example, another PR currently open has the [same missing param description for `swagger` in ScalaGatlingCodegen.java](https://github.com/viclovsky/swagger-codegen/blob/a270bbc1c9b67cb5876948ac70f7d1f8f6ee0a89/modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/ScalaGatlingCodegen.java#L199), but that PR's [AppVeyor build](https://ci.appveyor.com/project/WilliamCheng/swagger-codegen-wh2wu/build/master-7584) is all green. nbd, just weird that it would fail on this PR and not the other.

I've made those changes locally and run the javadocs:

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] swagger-codegen-project ............................ SUCCESS [ 12.589 s]
[INFO] swagger-codegen (core library) ..................... SUCCESS [  0.773 s]
[INFO] swagger-codegen (executable) ....................... SUCCESS [  0.835 s]
[INFO] swagger-codegen (maven-plugin) ..................... SUCCESS [  1.383 s]
[INFO] swagger-generator .................................. SUCCESS [  0.313 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 16.341 s
[INFO] Finished at: 2018-01-25T13:10:38-05:00
[INFO] Final Memory: 34M/478M
[INFO] ------------------------------------------------------------------------
```

**Pull request merged**